gap_ninafw.go

```diff
@@ -0,0 +1,226 @@
+//go:build ninafw
+
+package bluetooth
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	ErrConnect = errors.New("could not connect")
+)
+
+// Scan starts a BLE scan.
+func (a *Adapter) Scan(callback func(*Adapter, ScanResult)) error {
+	if a.scanning {
+		return errScanning
+	}
+
+	if err := a.hci.leSetScanEnable(false, true); err != nil {
+		return err
+	}
+
+	// passive scanning, every 40ms, for 30ms
+	if err := a.hci.leSetScanParameters(0x00, 0x0080, 0x0030, 0x00, 0x00); err != nil {
+		return err
+	}
+
+	a.scanning = true
+
+	// scan with duplicates
+	if err := a.hci.leSetScanEnable(true, false); err != nil {
+		return err
+	}
+
+	lastUpdate := time.Now().UnixNano()
+
+	for {
+		if err := a.hci.poll(); err != nil {
+			return err
+		}
+
+		switch {
+		case a.hci.advData.reported:
+			adf := AdvertisementFields{}
+			if a.hci.advData.eirLength > 31 {
+				if _debug {
+					println("eirLength too long")
+				}
+
+				a.hci.clearAdvData()
+				continue
+			}
+
+			for i := 0; i < int(a.hci.advData.eirLength); {
+				l, t := int(a.hci.advData.eirData[i]), a.hci.advData.eirData[i+1]
+				if l < 1 {
+					break
+				}
+
+				switch t {
+				case 0x02, 0x03:
+					// 16-bit Service Class UUID
+				case 0x06, 0x07:
+					// 128-bit Service Class UUID
+				case 0x08, 0x09:
+					if _debug {
+						println("local name", string(a.hci.advData.eirData[i+2:i+1+l]))
+					}
+
+					adf.LocalName = string(a.hci.advData.eirData[i+2 : i+1+l])
+				case 0xFF:
+					// Manufacturer Specific Data
+				}
+
+				i += l + 1
+			}
+
+			callback(a, ScanResult{
+				Address: Address{MACAddress{MAC: makeAddress(a.hci.advData.peerBdaddr)},
+					a.hci.advData.peerBdaddrType},
+				RSSI: int16(a.hci.advData.rssi),
+				AdvertisementPayload: &advertisementFields{
+					AdvertisementFields: adf,
+				},
+			})
+
+			a.hci.clearAdvData()
+			time.Sleep(10 * time.Millisecond)
+
+		default:
+			if !a.scanning {
+				return nil
+			}
+
+			if _debug && (time.Now().UnixNano()-lastUpdate)/int64(time.Second) > 1 {
+				println("still scanning...")
+				lastUpdate = time.Now().UnixNano()
+			}
+
+			time.Sleep(10 * time.Millisecond)
+		}
+	}
+
+	return nil
+}
+
+func (a *Adapter) StopScan() error {
+	if !a.scanning {
+		return errNotScanning
+	}
+
+	if err := a.hci.leSetScanEnable(false, false); err != nil {
+		return err
+	}
+
+	a.scanning = false
+
+	return nil
+}
+
+// Address contains a Bluetooth MAC address.
+type Address struct {
+	MACAddress
+
+	random uint8
```

Comment:
How is this different from `isRandom` that's part of `MACAddress`?

Reply:
Removed.